File: src/domain/notification-policy.ts
import type { SurfaceState } from "./surface-state";
import type { AppSettings } from "./settings";
import type { TaskItem } from "./task-item";

export type PanelLayout = "collapsed" | "peek" | "expanded";

export interface NotificationDecision {
  layout: PanelLayout;
  autoHideMs: number | null;
  peek: boolean;
}

/**
 * The layout to return to when the user dismisses an expanded task list.
 * Auto-hiding notifications are transient, so unread completion state must
 * not pin the panel open as a peek forever.
 */
export function notificationRestingLayout(decision: NotificationDecision): PanelLayout {
  return decision.peek && decision.autoHideMs === null ? "peek" : "collapsed";
}

/** True when an update removes the final running task from the snapshot. */
export function didFinishLastRunning(
  previous: readonly Pick<TaskItem, "status">[],
  next: readonly Pick<TaskItem, "status">[],
): boolean {
  return (
    previous.some((task) => task.status === "running") &&
    !next.some((task) => task.status === "running")
  );
}

export function isMuted(settings: Pick<AppSettings, "mutedUntil">, now = Date.now()): boolean {
  if (!settings.mutedUntil) return false;
  const until = Date.parse(settings.mutedUntil);
  return Number.isFinite(until) && until > now;
}

/**
 * Environment-style notifications.
 * working/waiting/failed: peek and stay. completed: peek ~5s.
 */
export function decideNotification(
  state: SurfaceState,
  options: { muted?: boolean; focusMode?: boolean } = {},
): NotificationDecision {
  if (options.muted) {
    return { layout: "collapsed", autoHideMs: null, peek: false };
  }

  switch (state.kind) {
    case "waiting":
      return { layout: "peek", autoHideMs: null, peek: true };
    case "failed":
      return { layout: "peek", autoHideMs: null, peek: true };
    case "completed":
      if (options.focusMode) {
        return { layout: "collapsed", autoHideMs: null, peek: false };
      }
      return { layout: "peek", autoHideMs: 5000, peek: true };
    case "working":
      return { layout: "peek", autoHideMs: null, peek: true };
    case "idle":
      return { layout: "collapsed", autoHideMs: null, peek: false };
  }
}

File: src-tauri/src/domain/notification_policy.rs
//! Notification policy. Keep in sync with src/domain/notification-policy.ts
#![allow(dead_code)]

use serde::{Deserialize, Serialize};

use super::settings::AppSettings;
use super::surface_state::SurfaceState;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum PanelLayout {
    Collapsed,
    Peek,
    Expanded,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NotificationDecision {
    pub layout: PanelLayout,
    pub auto_hide_ms: Option<u64>,
    pub peek: bool,
}

/// Keep in sync with src/domain/notification-policy.ts
pub fn decide_notification(state: &SurfaceState, settings: &AppSettings) -> NotificationDecision {
    if muted(settings) {
        return NotificationDecision {
            layout: PanelLayout::Collapsed,
            auto_hide_ms: None,
            peek: false,
        };
    }

    match state {
        SurfaceState::Waiting { .. } => NotificationDecision {
            layout: PanelLayout::Peek,
            auto_hide_ms: None,
            peek: true,
        },
        SurfaceState::Failed { .. } => NotificationDecision {
            layout: PanelLayout::Peek,
            auto_hide_ms: None,
            peek: true,
        },
        SurfaceState::Completed { .. } => {
            if settings.focus_mode {
                NotificationDecision {
                    layout: PanelLayout::Collapsed,
                    auto_hide_ms: None,
                    peek: false,
                }
            } else {
                NotificationDecision {
                    layout: PanelLayout::Peek,
                    auto_hide_ms: Some(5000),
                    peek: true,
                }
            }
        }
        SurfaceState::Working { .. } => NotificationDecision {
            layout: PanelLayout::Peek,
            auto_hide_ms: None,
            peek: true,
        },
        SurfaceState::Idle => NotificationDecision {
            layout: PanelLayout::Collapsed,
            auto_hide_ms: None,
            peek: false,
        },
    }
}

fn muted(settings: &AppSettings) -> bool {
    let Some(until) = settings.muted_until.as_deref() else {
        return false;
    };
    match chrono::DateTime::parse_from_rfc3339(until) {
        Ok(deadline) => deadline > chrono::Utc::now(),
        Err(_) => false,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::domain::{TaskItem, TaskSource, TaskStatus};

    fn waiting() -> SurfaceState {
        SurfaceState::Waiting {
            task: TaskItem {
                id: "t".into(),
                source: TaskSource::Codex,
                title: "ask".into(),
                summary: None,
                status: TaskStatus::Waiting,
                started_at: None,
                updated_at: "2026-08-13T04:00:00.000Z".into(),
                completed_at: None,
                unread: true,
                action: None,
            },
        }
    }

    fn working() -> SurfaceState {
        let SurfaceState::Waiting { mut task } = waiting() else {
            unreachable!()
        };
        task.status = TaskStatus::Running;
        task.unread = false;
        SurfaceState::Working { task }
    }

    #[test]
    fn muted_never_peeks() {
        let mut settings = AppSettings::default();
        settings.muted_until = Some("2099-01-01T00:00:00.000Z".into());
        let decision = decide_notification(&waiting(), &settings);
        assert!(!decision.peek);
        assert_eq!(decision.layout, PanelLayout::Collapsed);
    }

    #[test]
    fn working_stays_visible() {
        let decision = decide_notification(&working(), &AppSettings::default());
        assert!(decision.peek);
        assert_eq!(decision.layout, PanelLayout::Peek);
        assert_eq!(decision.auto_hide_ms, None);
    }
}
